README.pod
=pod

=head1 NAME

Mojolicious::Plugin::DigestAuth - HTTP Digest Authentication for Mojolicious

=head1 SYNOPSIS

   $self->plugin('digest_auth');

   # In your action
   return unless $self->digest_auth(allow => { sshaw => 'password' });

   # Or, in startup()
   my $r = $self->digest_auth('/admin', allow => { sshaw => 'password' });
   $r->route('/new')->to('users#new');

=head1 CONFIGURATION

=head2 SETUP

Configuration can be done globally when loading the plugin

    $self->plugin('digest_auth', %options)

or locally when calling L<< C<digest_auth>|/digest_auth >>

    $self->digest_auth(%options);

Local options override their global counterparts. For example, the following
will apply to all authentication requests

   # setup()
   $self->plugin('digest_auth', realm   => 'Thangz',
                                expires => 120,
                                allow   => '/path/to/htdigest_file');


   # controller
   sub show
   {
       my $self = shift;
       return unless $self->digest_auth;

       # ...
   }

But can be overridden within an action

   sub edit
   {
       my $self = shift;
       return unless $self->digest_auth(realm   => 'RealmX',
                                        expires => 24*3600,
                                        allow   => { sshaw => 'Ay3Br4h_!' });
       # ...
   }

For a full list of options see L</digest_auth>.

=head2 AUTHENTICATION

By default MD5/auth authentication is performed. This is configurable, see L</digest_auth>.

=head3 DB

Authentication information is given via the C<allow> option and can be retrieved
from a variety of sources:

=over 4

=item * A hash reference without a realm

    $self->plugin('digest_auth', allow => { sshaw => 'my_pAzw3rD',
                                            admin => '->fofinha!' });

In this case users will either be placed into the realm given by the C<realm> option or
the default realm, C<WWW>.

Passwords must be given in plain text.

=item * A hash reference with realm(s)

    $self->plugin('digest_auth', allow => { 'Admin Realm' => { sshaw => 'my_pAzw3rD' },
                                            'WWW Users'   => { tony  => 'vrooooooom' });

Passwords must be given in plain text.

=item * A htdigest style file

    $self->plugin('digest_auth', allow => '/home/sshaw/www_users');

=item * An object with a C<get()> method that returns B<hashed> passwords

    $self->plugin('digest_auth', allow => $db);

Arguments are passed to C<get()> in the following order: C<realm, username>.

=back

=head3 PERFORMING AUTHENTICATION

Authentication can be performed by calling the C<digest_auth> method
from within the action you'd like to protect:

   sub some_action
   {
       my $self = shift;
       return unless $self->digest_auth;

       # Authenticated users get here
   }

If authentication is successful C<digest_auth> returns true, otherwise C<undef> is returned
and a HTTP 401 status code and the message: C<HTTP 401: Unauthorized> are sent
to the client. Currently this message cannot be changed.

Authentication can also be performed for a set of routes by calling
C<digest_auth> from within your application's startup function. This form performs authentication automatically
for all of the routes defined under the given URL:

   package YourWebApp;

   use Mojo::Base 'Mojolicious';

   sub startup
   {
     my $self = shift;
     $self->plugin('digest_auth', %options);

     # ...

     my $admin = $self->digest_auth('/admin');
     $admin->route('/new')->to('users#new');
     $admin->route('/edit/:id')->to('users#edit');
   }

In this case authentication is performed via a L<bridge|Mojolicious::Guides::Routing/Bridges> with a callback.

=head3 WEB SERVERS

Authentication will fail if your application is sitting behind a web server does not pass the Authorization header 
to your application. In Apache this can be achieved with C<mod_rewrite>:

   RewriteEngine On
   RewriteRule ^ - [E=X-HTTP_AUTHORIZATION:%{HTTP:Authorization}]

=head1 METHODS

=head2 plugin

     $self->plugin('digest_auth', %options)

Loads the plugin and sets up the defaults given by C<%options>.

=head3 Arguments

C<%options>

See L</digest_auth>.

=head3 Errors

This method will C<croak> if if any of the options are invalid or if there is an error loading the password database.

=head2 digest_auth

     $self->digest_auth(%options)
     $routes = $self->digest_auth($url, %options)

=head3 Arguments

C<$url>

Optional. If provided authentication will be performed for all routes defined under C<$url>.
See L</PERFORMING AUTHENTICATION>.

C<%options>

=over 4

=item * C<< allow => { user => password } >>

=item * C<< allow => { realm => { user => password }} >>

=item * C<< allow => 'htdigest_file' >>

=item * C<< allow => $obj >>

See L</DB>.

=item * C<< algorithm => 'MD5' | 'MD5-sess' >>

Digest algorithm, either C<'MD5'> or C<'MD5-sess'>. Defaults to C<'MD5'>, C<'MD5-sess'> requires a C<qop>.

=item * C<< domain => '/path' | 'your.domain.com' >>

Authentication domain. Defaults to C<'/'>.

=item * C<< expires => seconds >>

Nonce lifetime. Defaults to C<300> seconds (5 minutes).

=item * C<< qop => 'auth' | '' >>

Quality of protection. Defaults to C<'auth'>.  C<auth-int> is not supported.

=item * C<< realm => 'Your Realm' >>

Authentication realm. Defaults to C<'WWW'>.

=item * C<< support_broken_browsers => 1 | 0 >>

When processing requests from certain browsers skip steps that would otherwise result in a HTTP 400 response. Defaults to C<1>. 

Currently only applies to IE 5 and 6. These two browsers fail to append the query string to the URI included in the 
Authorization header and, after authenticating, fail to include the opaque value. 

=back

=head3 Returns

Without a URL prefix:

True if authentication was successful, C<undef> otherwise. If unsuccessful a HTTP 401 status code and message are sent to the client.

With a URL prefix:

An instance of L<Mojolicious::Routes>. See L</PERFORMING AUTHENTICATION>.

=head3 Errors

Will C<croak> if any of the options are invalid.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Plugin::BasicAuth>, http://en.wikipedia.org/wiki/Digest_access_authentication

=head1 AUTHOR

Skye Shaw (sshaw AT lucas.cis.temple.edu)

=head1 LICENSE

Copyright (c) 2011 Skye Shaw.
This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.


CHANGES
-------------------------
v0.03	2011-11-24
-------------------------
* Fixed URL comparison (Bug #1) 
* Fixed various things broken by new releases of Mojolicious
* Fixed uninitialized value warnings from checksum()
* Changed how the current request's URL is determined
* Added check for X-HTTP_AUTHORIZATION
* Added option 'support_broken_browsers'


t/TestHelper.pm
package TestHelper;

use strict;
use warnings;
use base 'Exporter';

use Mojolicious::Lite;
use Mojolicious::Plugin::DigestAuth::Util qw{checksum parse_header};

our @EXPORT = qw{build_auth_request users create_action IE6};

my $users = { sshaw => 'itzme!' };
sub users { $_[0] ? $users->{$_[0]} : $users }
sub IE6 { 'Mozilla/5.0 (compatible; MSIE 6.0; Windows NT 5.1)' }

sub create_action
{
    my $options = { @_ };
    $options->{allow} ||= users();

    my $env = delete $options->{env} || {};
    sub { 
	my $self = shift;
	$self->app->plugin('digest_auth', $options);
	$self->req->env($env);
	$self->render_text("You're in!") if $self->digest_auth;
    };
}

# This fx() should use the same code as DigestAuth!
sub build_auth_request
{
    my ($tx, %defaults) = @_;
    my $req_header = parse_header($tx->res->headers->www_authenticate);	
    my $res_header = {};
    my $user = delete $defaults{username};
    my $pass = delete $defaults{password};
    my @common_parts = qw{algorithm nonce opaque realm};	

    $user = 'sshaw' if !defined $user;
    $pass = users($user) || '' if !defined $pass;

    local $_;
    @$res_header{@common_parts, keys %defaults} = (@$req_header{@common_parts}, values %defaults);

    # Test::Mojo handles the url differently between versions
    # What versions? Is this still necessary, maybe it was pre 1.32?
    if(!defined $res_header->{uri}) {
	$res_header->{uri} = $tx->req->url->path->to_string;
	$res_header->{uri} .= '?' . $tx->req->url->query if $tx->req->url->query->to_string;
    }
    
    $res_header->{nc} ||= 1;
    $res_header->{cnonce} ||= time();
    $res_header->{qop} ||= 'auth';
    $res_header->{username} = $user;
    $res_header->{response} = checksum(checksum($user, $res_header->{realm}, $pass),
				       $res_header->{nonce},
				       $res_header->{nc},
				       $res_header->{cnonce},
				       $res_header->{qop},
				       checksum($tx->req->method, $res_header->{uri}));    

    { Authorization => sprintf('Digest %s', join ', ', map { qq|$_="$res_header->{$_}"| } keys %$res_header) };      
}

1;


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
     NAME          => 'Mojolicious::Plugin::DigestAuth',
     AUTHOR        => 'Skye Shaw <sshaw AT lucas.cis.temple.edu>',	      
     PREREQ_PM     => { 'Mojolicious' => '1.31' },
     LICENSE       => 'perl',
     VERSION_FROM  => 'lib/Mojolicious/Plugin/DigestAuth.pm',
     ABSTRACT_FROM => 'lib/Mojolicious/Plugin/DigestAuth.pm',
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ?
       (META_MERGE  => { resources => { bugtracker => 'http://github.com/sshaw/Mojolicious-Plugin-DigestAuth/issues' } }) : ())     
);

# Create README.pod for a repo's GitHub page. Unlike CPAN, GitHub won't 
# display the module's POD, it looks for a README.*
sub MY::postamble
{
    my $self = shift;
    return if -r 'README' or ! -r $self->{VERSION_FROM};
    return<<END_MAKE;
README.pod: $self->{VERSION_FROM}
	\@perldoc -uT $self->{VERSION_FROM} > README.pod
END_MAKE
}


lib/Mojolicious/Plugin/DigestAuth/DB.pm
package Mojolicious::Plugin::DigestAuth::DB;

use strict;
use warnings;

use Carp ();

sub get
{
    Carp::croak 'usage: ', __PACKAGE__, '->get(REALM, USER)' unless @_ == 3; 
    my($self, $realm, $user) = @_;    
    return unless defined $realm and defined $user;
    $self->{users}->{$realm}->{$user};
}

package Mojolicious::Plugin::DigestAuth::DB::File;

use base 'Mojolicious::Plugin::DigestAuth::DB';

sub new
{
    my ($class, $file) = @_;
    Carp::croak 'usage: ', __PACKAGE__, '->new(FILE)' unless $file;
   
    local $_;
    my $users = {};

    open my $passwords, '<', $file or Carp::croak "error opening digest file '$file': $!";

    while(<$passwords>) {
	# "\r\n" used w/ Apache digest files on Win?
	chomp;
	next unless $_;
	
	# user:realm:hashed_password
	my @user = split /:/, $_, 3;
	if(@user != 3 || grep { !defined or !length } @user) {
	    Carp::croak "password file '$file' contains an invalid entry: $_";
	}
	
	$users->{$user[1]}->{$user[0]} = $user[2];	    
    }

    bless { users => $users }, $class;
}

package Mojolicious::Plugin::DigestAuth::DB::Hash;

use base 'Mojolicious::Plugin::DigestAuth::DB';
use Mojolicious::Plugin::DigestAuth::Util 'checksum';

sub new
{
    my ($class, $config) = @_;
    Carp::croak 'usage: ', __PACKAGE__, '->new(HASH)' unless $config and ref $config eq 'HASH';

    my $users;
    for my $realm (keys %$config) {
	Carp::croak "config for realm '$realm' is invalid: values must be a HASH" unless ref $config->{$realm} eq 'HASH';
	while(my ($user, $password) = each %{$config->{$realm}}) {
	    #allow blank passowrdz?
	    $users->{$realm}->{$user} = checksum($user, $realm, $password);
	}
    }

    bless { users => $users }, $class;
}

1;


lib/Mojolicious/Plugin/DigestAuth/Util.pm
package Mojolicious::Plugin::DigestAuth::Util;

use strict;
use warnings;

use Mojo::Util qw{md5_sum};
use base 'Exporter';

our @EXPORT_OK = qw{checksum parse_header quote unquote b64_encode b64_decode};

sub checksum
{
  local $_;
  md5_sum join ':', grep(defined, @_);
}

sub parse_header
{
    my $header = shift;
    my $parsed;

    # TODO: I think there's a browser with a quoting issue that might affect this
    if($header && $header =~ s/^Digest\s//) {
        while($header =~ /([a-zA-Z]+)=(".*?"|[^,]+)/g){
	  $parsed->{$1} = unquote($2);
        }
    }

    $parsed;
}

#
# These 4 functions are used to maintain backwards compatibility with older versions
# of Mojolicious and will be removed in the next release.
#
sub quote
{
  my $str = shift;
  my $t = Mojo::Util::quote($str);
  $str = $t if $t;
  $str;
}

sub unquote
{
  my $str = shift;
  my $t = Mojo::Util::unquote($str);
  $str = $t if $t;
  $str;
}

sub b64_encode
{
  my $str = shift;
  my $t = Mojo::Util::b64_encode($str);
  $str = $t if $t;
  $str;
}

sub b64_decode
{
  my $str = shift;
  my $t = Mojo::Util::b64_decode($str);
  $str = $t if $t;
  $str;
}

1;


lib/Mojolicious/Plugin/DigestAuth/RequestHandler.pm
package Mojolicious::Plugin::DigestAuth::RequestHandler;

use strict;
use warnings;

use Carp 'croak';
use Scalar::Util 'weaken';

use Mojolicious::Plugin::DigestAuth::Util qw{checksum parse_header quote b64_encode b64_decode};

my $QOP_AUTH = 'auth';
my $QOP_AUTH_INT = 'auth-int';
my %VALID_QOPS = ($QOP_AUTH => 1); #, $QOP_AUTH_INT => 1);

my $ALGORITHM_MD5 = 'MD5';
my $ALGORITHM_MD5_SESS = 'MD5-sess';
my %VALID_ALGORITHMS = ($ALGORITHM_MD5 => 1, $ALGORITHM_MD5_SESS => 1);

sub new
{
    my ($class, $config) = @_;
    my $header = {
        qop	  => $config->{qop},
        realm     => $config->{realm}     || '',
        domain    => $config->{domain}    || '/',
        algorithm => $config->{algorithm} || $ALGORITHM_MD5,
    };

    # No qop = ''
    $header->{qop} = $QOP_AUTH unless defined $header->{qop}; # "$QOP_AUTH,$QOP_AUTH_INT" 
    $header->{opaque} = checksum($header->{domain}, $config->{secret});

    my $self = {
	qops           => {},
        opaque         => $header->{opaque},
        secret         => $config->{secret},
        expires        => $config->{expires},
	algorithm      => $header->{algorithm},
        password_db    => $config->{password_db},
        default_header => $header,
	support_broken_browsers => $config->{support_broken_browsers}
    };

    $self->{support_broken_browsers} = 1 unless defined $self->{support_broken_browsers};

    for my $qop (split /\s*,\s*/, $header->{qop}) {
        croak "unsupported qop: $qop" unless $VALID_QOPS{$qop};
        $self->{qops}->{$qop} = 1;
    }

    croak "unsupported algorithm: $self->{algorithm}" unless $VALID_ALGORITHMS{$self->{algorithm}};
    croak "algorithm $ALGORITHM_MD5_SESS requires a qop" if $self->{algorithm} eq $ALGORITHM_MD5_SESS and ! %{$self->{qops}};

    bless $self, $class;
}  

sub _request
{
    (shift)->_controller->req;
}

sub _response
{
    (shift)->_controller->res;
}

sub _controller
{
    (shift)->{controller};
}

sub _nonce_expired
{
    my ($self, $nonce) = @_;
    my $t;

    $t = ($self->_parse_nonce($nonce))[0];
    $t && (time() - int($t)) > $self->{expires};
}

sub _parse_nonce
{
    my ($self, $nonce) = @_;
    split ' ', b64_decode($nonce), 2;
}

sub _valid_nonce
{
    my ($self, $nonce) = @_;
    my ($t, $sig) = $self->_parse_nonce($nonce);

    $t && $sig && $sig eq checksum($t, $self->{secret});
}

sub _create_nonce
{
    my $self  = shift;
    my $t     = time();
    my $nonce = b64_encode(sprintf('%s %s', $t, checksum($t, $self->{secret})));
    chomp $nonce;
    $nonce;
}

sub authenticate
{
    my $self = shift;

    $self->{controller} = shift;
    weaken $self->{controller};

    $self->{response_header} = { %{$self->{default_header}} };

    my $auth = $self->_auth_header;
    if($auth) {
        my $header = parse_header($auth);
        if(!$self->_valid_header($header)) {
            $self->_bad_request;
            return;
        }

        if($self->_authorized($header)) {
            return 1 unless $self->_nonce_expired($header->{nonce});
            $self->{response_header}->{stale} = 'true';
        }
    }

    $self->_unauthorized;
}

# TODO: $self->_request->headers->proxy_authorization
sub _auth_header
{
  my $self = shift;
  $self->_request->headers->authorization or  
  $self->_request->env->{'X_HTTP_AUTHORIZATION'} # Mojo does s/-/_/g
}

sub _unauthorized
{
    my $self = shift;
    my $header = $self->_build_auth_header;

    $self->_response->headers->www_authenticate($header);
    $self->_response->code(401);
    $self->_controller->render(text => 'HTTP 401: Unauthorized');
}

sub _bad_request
{
    my $self = shift;
    $self->_response->code(400);
    $self->_controller->render(text => 'HTTP 400: Bad Request');
}

sub _valid_header
{
    my ($self, $header) = @_;

    $self->_header_complete($header) &&
    $self->_url_matches($header->{uri}) &&
    $self->_valid_qop($header->{qop}, $header->{nc}) &&
    $self->_valid_opaque($header->{opaque}) &&
    $self->{algorithm} eq $header->{algorithm};
}

sub _url_matches
{
    my $self = shift;

    my $auth_url = shift;
    return unless $auth_url;
    $auth_url = _normalize_url($auth_url);
    
    my $req_url = $self->_url; 
    
    if($self->_support_broken_browser) {
      # IE 5/6 do not append the querystring on GET requests
      my $i = index($req_url, '?');
      if($self->_request->method eq 'GET' && $i != -1 && index($auth_url, '?') == -1) {
      	  $auth_url .= '?' . substr($req_url, $i+1);
      }
    }
        
    $auth_url eq $req_url;
}

#
# We try to avoid using the URL provided by Mojo because:
#
# * Depending on the app's config it will not contain the URL requested by the client
#   it will contain PATH_INFO + QUERY_STRING i.e. /mojo.pl/users/sshaw?x=y will be /users/sshaw?x=y
#    
# * Mojo::URL has/had several bugs and has undergone several changes that have broken backwards 
#   compatibility.
#
sub _url
{
  my $self = shift;
  my $env = $self->_request->env;
  my $url;

  if($env->{REQUEST_URI}) {
    $url = $env->{REQUEST_URI};
  }
  elsif($env->{SCRIPT_NAME}) {
    $url = $env->{SCRIPT_NAME};
    $url .= $env->{PATH_INFO} if $env->{PATH_INFO};
    $url .= "?$env->{QUERY_STRING}" if $env->{QUERY_STRING};
  }
  elsif($self->_request->url) {
    $url = $self->_request->url->to_string;
  }
  else {
    $url = '/';
  }

  _normalize_url($url);
}

# We want the URL to be relative to '/'
sub _normalize_url
{
  my $s = shift;
  $s =~ s|^https?://[^/?#]*||i;    
  $s =~ s|/{2,}|/|g;    

  my $url = Mojo::URL->new($s);  
  my @parts = @{$url->path->parts};
  my @normalized;
  
  for my $part (@parts) {
    if($part eq '..' && @normalized) {
      pop @normalized;
      next;
    }
    
    push @normalized, $part; 
  }

  $url->path->parts(\@normalized);
  $url->path->leading_slash(0);
  $url->to_string;
}

# TODO (maybe): IE 6 sends a new nonce every time when using MD5-sess
sub _support_broken_browser
{
    my $self = shift;
    $self->{support_broken_browsers} && $self->_request->headers->user_agent =~ m|\bMSIE\s+[56]\.|;    
}

sub _valid_qop
{
  my ($self, $qop, $nc) = @_;
  my $valid;

  #
  # Either there's no QOP from the client and we require one, or the client does not
  # send a qop because they dont support what we want (e.g., auth-int). 
  #
  # And, if there's a qop, then there must be a nonce count.
  #
  if(defined $qop) {
    $valid = $self->{qops}->{$qop} && $nc;
  }
  else {
    $valid = !%{$self->{qops}} && !defined $nc;
  }
  
  $valid;
}

sub _valid_opaque
{
  my ($self, $opaque) = @_;

  # IE 5 & 6 only sends opaque with the initial reply but we'll just ignore it regardless
  $self->_support_broken_browser || $opaque && $opaque eq $self->{opaque};  
}

sub _header_complete
{
    my ($self, $header) = @_;
    
    $header &&
    $header->{realm} &&
    $header->{nonce} &&
    $header->{response} &&
    $header->{algorithm} && 
    exists $header->{username};
}

sub _build_auth_header
{
    my $self   = shift;
    my $header = $self->{response_header};

    if($header->{stale} || !$header->{nonce}) {
        $header->{nonce} = $self->_create_nonce;
    }

    my %no_quote;
    @no_quote{qw{algorithm stale}} = ();

    my @auth;
    while(my ($k, $v) = each %$header) {
      next unless $v;
      $v = quote($v) unless exists $no_quote{$k};
      push @auth, "$k=$v";
    }

    'Digest ' . join(', ', @auth);
}

sub _authorized
{
    my ($self, $header) = @_;
    return unless $self->_valid_nonce($header->{nonce});

    my $a1 = $self->_compute_a1($header);
    return unless $a1;

    my @fields = ($a1, $header->{nonce});
    if($header->{qop}) {
        push @fields, $header->{nc},
                      $header->{cnonce},
                      $header->{qop},
                      $self->_compute_a2($header);
    }
    else {
        push @fields, $self->_compute_a2($header);
    }

    checksum(@fields) eq $header->{response};
}

sub _compute_a1
{
    my ($self, $header) = @_;
    my $hash = $self->{password_db}->get($header->{realm}, $header->{username});

    if($hash && $header->{algorithm} && $header->{algorithm} eq $ALGORITHM_MD5_SESS) {
        $hash = checksum($hash, $header->{nonce}, $header->{cnonce});
    }

    $hash;
}

sub _compute_a2
{
    my ($self, $header) = @_;
    my @fields = ($self->_request->method, $header->{uri});

# Not yet...
#     if(defined $header->{qop} && $header->{qop} eq $QOP_AUTH_INT) {
#         # TODO: has body been decoded?
#       push @fields, checksum($self->_request->content->headers->to_string . "\015\012\015\012" . $self->_request->body);
#     }

    checksum(@fields);
}

1;


lib/Mojolicious/Plugin/DigestAuth.pm
package Mojolicious::Plugin::DigestAuth;

use strict;
use warnings;

use Carp 'croak';
use Scalar::Util 'blessed';

use Mojo::Base 'Mojolicious::Plugin';
use Mojolicious::Plugin::DigestAuth::DB;
use Mojolicious::Plugin::DigestAuth::RequestHandler;

our $VERSION = '0.03';

sub register
{
    my ($self, $app, $user_defaults) = @_;

    my %defaults = %$user_defaults;
    $defaults{realm}   ||= 'WWW';
    $defaults{secret}  ||= $app->secret;
    $defaults{expires} ||= 300;

    $app->helper(digest_auth => sub {
        my $c = shift;

        my $route;
        $route = shift if @_ % 2;

        my $options = { %defaults, @_ };
        croak 'you must setup an authentication source via the "allow" option' if !defined $options->{allow};

        my $allow = delete $options->{allow};
        if(blessed($allow) && $allow->can('get')) {
          $options->{password_db} = $allow;
        }
        elsif(ref($allow) eq 'HASH') {
          # Normalize simple config, otherwise we assume a hash of: realm => { user => 'password' ... }
          if(ref((values %$allow)[0]) ne 'HASH') {
            $allow = { $options->{realm} => { %$allow } };
          }

          $options->{password_db} = Mojolicious::Plugin::DigestAuth::DB::Hash->new($allow);
        }
        else {
          # Assume it's a file
          $options->{password_db} = Mojolicious::Plugin::DigestAuth::DB::File->new($allow);
        }

        my $handler = Mojolicious::Plugin::DigestAuth::RequestHandler->new($options);
        if($route) {
            return $c->app->routes->bridge($route)->to(cb => sub {
                $handler->authenticate(shift);
            });
        }

        $handler->authenticate($c);
    });
}

1;

__END__

=pod

=head1 NAME

Mojolicious::Plugin::DigestAuth - HTTP Digest Authentication for Mojolicious

=head1 SYNOPSIS

   $self->plugin('digest_auth');

   # In your action
   return unless $self->digest_auth(allow => { sshaw => 'password' });

   # Or, in startup()
   my $r = $self->digest_auth('/admin', allow => { sshaw => 'password' });
   $r->route('/new')->to('users#new');

=head1 CONFIGURATION

=head2 SETUP

Configuration can be done globally when loading the plugin

    $self->plugin('digest_auth', %options)

or locally when calling L<< C<digest_auth>|/digest_auth >>

    $self->digest_auth(%options);

Local options override their global counterparts. For example, the following
will apply to all authentication requests

   # setup()
   $self->plugin('digest_auth', realm   => 'Thangz',
                                expires => 120,
                                allow   => '/path/to/htdigest_file');


   # controller
   sub show
   {
       my $self = shift;
       return unless $self->digest_auth;

       # ...
   }

But can be overridden within an action

   sub edit
   {
       my $self = shift;
       return unless $self->digest_auth(realm   => 'RealmX',
                                        expires => 24*3600,
                                        allow   => { sshaw => 'Ay3Br4h_!' });
       # ...
   }

For a full list of options see L</digest_auth>.

=head2 AUTHENTICATION

By default MD5/auth authentication is performed. This is configurable, see L</digest_auth>.

=head3 DB

Authentication information is given via the C<allow> option and can be retrieved
from a variety of sources:

=over 4

=item * A hash reference without a realm

    $self->plugin('digest_auth', allow => { sshaw => 'my_pAzw3rD',
                                            admin => '->fofinha!' });

In this case users will either be placed into the realm given by the C<realm> option or
the default realm, C<WWW>.

Passwords must be given in plain text.

=item * A hash reference with realm(s)

    $self->plugin('digest_auth', allow => { 'Admin Realm' => { sshaw => 'my_pAzw3rD' },
                                            'WWW Users'   => { tony  => 'vrooooooom' });

Passwords must be given in plain text.

=item * A htdigest style file

    $self->plugin('digest_auth', allow => '/home/sshaw/www_users');

=item * An object with a C<get()> method that returns B<hashed> passwords

    $self->plugin('digest_auth', allow => $db);

Arguments are passed to C<get()> in the following order: C<realm, username>.

=back

=head3 PERFORMING AUTHENTICATION

Authentication can be performed by calling the C<digest_auth> method
from within the action you'd like to protect:

   sub some_action
   {
       my $self = shift;
       return unless $self->digest_auth;

       # Authenticated users get here
   }

If authentication is successful C<digest_auth> returns true, otherwise C<undef> is returned
and a HTTP 401 status code and the message: C<HTTP 401: Unauthorized> are sent
to the client. Currently this message cannot be changed.

Authentication can also be performed for a set of routes by calling
C<digest_auth> from within your application's startup function. This form performs authentication automatically
for all of the routes defined under the given URL:

   package YourWebApp;

   use Mojo::Base 'Mojolicious';

   sub startup
   {
     my $self = shift;
     $self->plugin('digest_auth', %options);

     # ...

     my $admin = $self->digest_auth('/admin');
     $admin->route('/new')->to('users#new');
     $admin->route('/edit/:id')->to('users#edit');
   }

In this case authentication is performed via a L<bridge|Mojolicious::Guides::Routing/Bridges> with a callback.

=head3 WEB SERVERS

Authentication will fail if your application is sitting behind a web server does not pass the Authorization header 
to your application. In Apache this can be achieved with C<mod_rewrite>:

   RewriteEngine On
   RewriteRule ^ - [E=X-HTTP_AUTHORIZATION:%{HTTP:Authorization}]

=head1 METHODS

=head2 plugin

     $self->plugin('digest_auth', %options)

Loads the plugin and sets up the defaults given by C<%options>.

=head3 Arguments

C<%options>

See L</digest_auth>.

=head3 Errors

This method will C<croak> if if any of the options are invalid or if there is an error loading the password database.

=head2 digest_auth

     $self->digest_auth(%options)
     $routes = $self->digest_auth($url, %options)

=head3 Arguments

C<$url>

Optional. If provided authentication will be performed for all routes defined under C<$url>.
See L</PERFORMING AUTHENTICATION>.

C<%options>

=over 4

=item * C<< allow => { user => password } >>

=item * C<< allow => { realm => { user => password }} >>

=item * C<< allow => 'htdigest_file' >>

=item * C<< allow => $obj >>

See L</DB>.

=item * C<< algorithm => 'MD5' | 'MD5-sess' >>

Digest algorithm, either C<'MD5'> or C<'MD5-sess'>. Defaults to C<'MD5'>, C<'MD5-sess'> requires a C<qop>.

=item * C<< domain => '/path' | 'your.domain.com' >>

Authentication domain. Defaults to C<'/'>.

=item * C<< expires => seconds >>

Nonce lifetime. Defaults to C<300> seconds (5 minutes).

=item * C<< qop => 'auth' | '' >>

Quality of protection. Defaults to C<'auth'>.  C<auth-int> is not supported.

=item * C<< realm => 'Your Realm' >>

Authentication realm. Defaults to C<'WWW'>.

=item * C<< support_broken_browsers => 1 | 0 >>

When processing requests from certain browsers skip steps that would otherwise result in a HTTP 400 response. Defaults to C<1>. 

Currently only applies to IE 5 and 6. These two browsers fail to append the query string to the URI included in the 
Authorization header and, after authenticating, fail to include the opaque value. 

=back

=head3 Returns

Without a URL prefix:

True if authentication was successful, C<undef> otherwise. If unsuccessful a HTTP 401 status code and message are sent to the client.

With a URL prefix:

An instance of L<Mojolicious::Routes>. See L</PERFORMING AUTHENTICATION>.

=head3 Errors

Will C<croak> if any of the options are invalid.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Plugin::BasicAuth>, http://en.wikipedia.org/wiki/Digest_access_authentication

=head1 AUTHOR

Skye Shaw (sshaw AT lucas.cis.temple.edu)

=head1 LICENSE

Copyright (c) 2011 Skye Shaw.
This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.
